Various updates to the blackmagicprobe runner

fabiobaltieri opened the pull request:
Hi, ~~quick one~~ for the BMP runner to catch up with https://github.com/blackmagic-debug/blackmagic/pull/1056/commits/240099dde9fae2be6f50a0913f065736500d0727, they changed the command, updating the runner to send both the old and new ones, and take both version of the flags.

EDIT: added two patches to also improve the serial port guessing.

--

Recently the blackmagicprobe command for "connect under reset" has been updated from "connect_srst" to "connect_rst". Update the flag name in the west blackmagicprobe binding, keep the old one as well for compatibility with out of tree boards, and send both the old and new commands to GDB so that it works with both the old and new firmware. GDB is going to get a "command not supported by this target" for one of the two and proceed with the other one.

fabiobaltieri committed:
west: runners: blackmagicprobe: update reset on connect command

Recently the blackmagicprobe command for "connect under reset" has been
updated from "connect_srst" to "connect_rst". Update the flag name in
the west blackmagicprobe binding, keep the old one as well for
compatibility with out of tree boards, and send both the old and new
commands to GDB so that it works with both the old and new firmware. GDB
is going to get a "command not supported by this target" for one of the
two and proceed with the other one.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

nordicjm commented on boards/arm/lora_e5_dev_board/board.cmake:
```diff
@@ -5,7 +5,7 @@ board_runner_args(pyocd "--flash-opt=-O reset_type=hw")
 board_runner_args(jlink "--device=STM32WLE5JC" "--speed=4000" "--reset-after-load")
 board_runner_args(stm32cubeprogrammer "--port=swd" "--reset-mode=hw")
-board_runner_args(blackmagicprobe "--gdb-serial=/dev/ttyBmpGdb" "--connect-srst")
+board_runner_args(blackmagicprobe "--gdb-serial=/dev/ttyBmpGdb" "--connect-rst")
```
Unrelated to this PR but why is this forcing a specific serial port name through a udev rule?

fabiobaltieri replied:
ttyBmpGdb is the the alias used in the official documentation and udev rules (https://github.com/blackmagic-debug/blackmagic/blob/main/driver/99-blackmagic-plugdev.rules) it's used by a few boards, ideally the runner would try to see if that exists automatically and then guess something else, that could be a valid followup.

nordicjm replied:
It is very limited, you can only have one programmer connected to a PC at any one time for example and requires that users have installed udev rules, which I can't see linked in the documentation. It also completely breaks support for working on windows

nordicjm replied:
I've opened an RFC on this issue: https://github.com/zephyrproject-rtos/zephyr/issues/53859

fabiobaltieri replied:
Yeah well, right now the fallback is just `/dev/ttyACM0` so unless I missed something windows (or mac) support and multiple programmers are already broken. But yeah agreed that the runner should have some smartness and these per board overrides should go, thanks for opening the RFC.

fabiobaltieri replied:
Added a couple patches to address the flag as well. Broadening the scope of the PR but since it hasn't been reviewed yet I've figured I'll just expand it.

fabiobaltieri committed:
west: runners: blackmagicprobe: guess the gdb port based on the OS

Add some logic in blackmagicprobe.py to guess the GDB serial port based
on the OS instead of just trying the flag or /dev/ttyACM0.

The new logic explicitly looks for the well known path for BMP on Linux
(/dev/ttyBmpGdb), so that we can remove it from the board files that
have it set explicitly right now. At the same time it still allows
setting the port from the flag and also checks an environment variable,
so that it can be used for systems with multiple probes.

As a fallback, it looks for a valid serial port, picking the first one
(the probe creates two, the GDB one is the first one), and raise an
exception if nothing is found.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri committed:
boards: drop explicit --gdb-serial

Now that the well known udev BMP port is used automatically, drop it
from the board files.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri commented:
Hey, since this had not been reviewed yet I added a couple patches to address https://github.com/zephyrproject-rtos/zephyr/issues/53859.

So patch 1 fixes the "connect under reset" bit
Patch 2 makes the runner port selection OS aware
Patch 3 drops the current explicit flags from board files

Note that I only tested this on Linux.

@nordicjm what do you think?

nordicjm commented on scripts/west_commands/runners/blackmagicprobe.py:
```diff
@@ -5,11 +5,59 @@
+        return sorted(ports)[0]
+
+    elif platform.startswith('win32'):
+        return 'COM0'
```
COM0 is not valid, starts at COM1, though you can get the current IDs from the registry somewhere but this is good enough for now

fabiobaltieri replied:
Woops, bumped to COM1 thanks. Yeah I looked up how to guess it and the best I found was https://stackoverflow.com/questions/12090503/listing-available-com-ports-with-python, don't like the idea of actually opening the port to see if it's there though. Happy to leave the smartness here for someone who actually uses it.

nordicjm reviewed (dismissed):
Looks good, thanks!

mbolivar-nordic commented on scripts/west_commands/runners/blackmagicprobe.py:
```diff
@@ -5,11 +5,59 @@
+LINUX_SERIAL_GLOB = '/dev/ttyACM*'
+DARWIN_SERIAL_GLOB = '/dev/cu.usbmodem*'
+
+def BlackMagicProbeGDBSerial(port):
```
camel case is for class names, not function names

fabiobaltieri replied:
Sorted.

mbolivar-nordic commented on scripts/west_commands/runners/blackmagicprobe.py:
```diff
@@ -5,11 +5,59 @@
+                    f'cannot find any valid port matching {DARWIN_SERIAL_GLOB}')
+        return sorted(ports)[0]
+
+    elif platform.startswith('win32'):
+        return 'COM1'
```
I get that there are some nice udev rules in the linux case, but it seems like this could be made better on Windows and macOS using pyserial.

```
In [1]: import serial.tools.list_ports

In [2]: serial.tools.list_ports.comports(include_links=True)
Out[2]: 
[<serial.tools.list_ports_linux.SysFS at 0x7fe5abb2c700>,
 <serial.tools.list_ports_linux.SysFS at 0x7fe5abb2e320>,
 <serial.tools.list_ports_linux.SysFS at 0x7fe5abb2c430>]

In [3]: [vars(info) for info in serial.tools.list_ports.comports(include_links=True)]
Out[3]: 
[{'device': '/dev/ttyS1',
  'name': 'ttyS1',
  'description': 'ttyS1',
  'hwid': 'PNP0501',
  'vid': None,
  'pid': None,
  'serial_number': None,
  'location': None,
  'manufacturer': None,
  'product': None,
  'interface': None,
  'usb_device_path': None,
  'device_path': '/sys/devices/pnp0/00:05',
  'subsystem': 'pnp',
  'usb_interface_path': None},
 {'device': '/dev/ttyS0',
  'name': 'ttyS0',
  'description': 'ttyS0',
  'hwid': 'PNP0501',
  'vid': None,
  'pid': None,
  'serial_number': None,
  'location': None,
  'manufacturer': None,
  'product': None,
  'interface': None,
  'usb_device_path': None,
  'device_path': '/sys/devices/pnp0/00:04',
  'subsystem': 'pnp',
  'usb_interface_path': None},
 {'device': '/dev/ttyACM0',
  'name': 'ttyACM0',
  'description': 'J-Link - CDC',
  'hwid': 'USB VID:PID=1366:0105 SER=000683441799 LOCATION=3-4.4.4:1.0',
  'vid': 4966,
  'pid': 261,
  'serial_number': '000683441799',
  'location': '3-4.4.4:1.0',
  'manufacturer': 'SEGGER',
  'product': 'J-Link',
  'interface': 'CDC',
  'usb_device_path': '/sys/devices/pci0000:20/0000:20:01.1/0000:21:00.0/0000:22:08.0/0000:2c:00.1/usb3/3-4/3-4.4/3-4.4.4',
  'device_path': '/sys/devices/pci0000:20/0000:20:01.1/0000:21:00.0/0000:22:08.0/0000:2c:00.1/usb3/3-4/3-4.4/3-4.4.4/3-4.4.4:1.0',
  'subsystem': 'usb',
  'usb_interface_path': '/sys/devices/pci0000:20/0000:20:01.1/0000:21:00.0/0000:22:08.0/0000:2c:00.1/usb3/3-4/3-4.4/3-4.4.4/3-4.4.4:1.0'}]
```

Please use the same `MISSING_REQUIREMENTS` pattern as jlink.py for the import, since pyserial is not part of the standard library.

fabiobaltieri replied:
I get this on osx:

```
>>> vars(serial.tools.list_ports.comports()[1])
{'device': '/dev/cu.usbmodem93BB879F1', 'name': 'cu.usbmodem93BB879F1', 'description': 'Black Magic Probe v1.9.0-rc1-40-g1252c55f-dirty', 'hwid': 'USB VID:PID=1D50:6018 SER=93BB879F LOCATION=0-1', 'vid': 7504, 'pid': 24600, 'serial_number': '93BB879F', 'location': '0-1', 'manufacturer': 'Black Magic Debug', 'product': 'Black Magic Probe v1.9.0-rc1-40-g1252c55f-dirty', 'interface': None}
>>> vars(serial.tools.list_ports.comports()[2])
{'device': '/dev/cu.usbmodem93BB879F3', 'name': 'cu.usbmodem93BB879F3', 'description': 'Black Magic Probe v1.9.0-rc1-40-g1252c55f-dirty', 'hwid': 'USB VID:PID=1D50:6018 SER=93BB879F LOCATION=0-1', 'vid': 7504, 'pid': 24600, 'serial_number': '93BB879F', 'location': '0-1', 'manufacturer': 'Black Magic Debug', 'product': 'Black Magic Probe v1.9.0-rc1-40-g1252c55f-dirty', 'interface': None}
```

No interface unfortunately but it may be ok to filter for product and get the first one. Waiting for an output on windows if you can.

fabiobaltieri replied:
actually don't bother, read the code (https://github.com/pyserial/pyserial/blob/31fa4807d73ed4eb9891a88a15817b439c4eea2d/serial/tools/list_ports_windows.py#L332-L334) I think I can rely on vid and pid being there, updated

mbolivar-nordic reviewed (changes requested):
Nice, thank you for keeping this runner alive. A couple of minor things, otherwise LGTM.

fabiobaltieri committed:
ci: west_cmds: add pyserial

Add the pyserial Python module so that it can be used for unit testing.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

mbolivar-nordic commented:
@fabiobaltieri here is the windows output for the same segger serial port, in case it's useful to you:

```
>>> [vars(info) for info in serial.tools.list_ports.comports(include_links=True)]
[{'device': 'COM3', 'name': 'COM3', 'description': 'Intel(R) Active Management Technology - SOL (COM3)', 'hwid': 'PCI\\VEN_8086&DEV_9D3D&SUBSYS_081C1028&REV_21\\3&11583659&0&B3', 'vid': None, 'pid': None, 'serial_number': None, 'location': None, 'manufacturer': 'Intel', 'product': None, 'interface': None}, {'device': 'COM4', 'name': 'COM4', 'description': 'JLink CDC UART Port (COM4)', 'hwid': 'USB VID:PID=1366:0105 SER=000683441799 LOCATION=1-9:x.0', 'vid': 4966, 'pid': 261, 'serial_number': '000683441799', 'location': '1-9:x.0', 'manufacturer': 'SEGGER', 'product': None, 'interface': None}]
```

fabiobaltieri commented:
> @fabiobaltieri here is the windows output for the same segger serial port, in case it's useful to you:
> 
> ```
> >>> [vars(info) for info in serial.tools.list_ports.comports(include_links=True)]
> [{'device': 'COM3', 'name': 'COM3', 'description': 'Intel(R) Active Management Technology - SOL (COM3)', 'hwid': 'PCI\\VEN_8086&DEV_9D3D&SUBSYS_081C1028&REV_21\\3&11583659&0&B3', 'vid': None, 'pid': None, 'serial_number': None, 'location': None, 'manufacturer': 'Intel', 'product': None, 'interface': None},
> 
> {'device': 'COM4', 'name': 'COM4', 'description': 'JLink CDC UART Port (COM4)', 'hwid': 'USB VID:PID=1366:0105 SER=000683441799 LOCATION=1-9:x.0', 'vid': 4966, 'pid': 261, 'serial_number': '000683441799', 'location': '1-9:x.0', 'manufacturer': 'SEGGER', 'product': None, 'interface': None}]
> ```
> 
> edit: separated entries in the results list above

Thanks, cool looks like vid and pid are indeed there for the USB device so this should work with the current version of this PR.

mbolivar-nordic reviewed (approved):
Excellent!

mbolivar-nordic commented:
It's "hilarious" that you had to go with 3 different ways to match this device. Sad that the interface descriptor isn't universally available.

fabiobaltieri commented:
> It's "hilarious" that you had to go with 3 different ways to match this device. Sad that the interface descriptor isn't universally available.

That would have been ideal. This reminded me of sigrok splitting libserialport out to its own library (https://github.com/sigrokproject/libserialport), similar story I guess.

TequinDragon commented:
I know this has been closed. However I'm getting 
```
COM15: No such file or directory.
"monitor" command not supported by this target.
Don't know how to attach.  Try "help target".
You can't do that when your target is `None'
The program is not being run.
```

May I propose that `blackmagicprobe_gdb_serial_win32()` return a COMXX port prefixed with `\\.\` so that GDB is instructed to open `\\.\COM15` for example. As this works on Windows 10.

I'm using Zephyr SDK `v0.15.2`, and Zephyr `v3.3.0-rc1`

fabiobaltieri commented:
> May I propose that `blackmagicprobe_gdb_serial_win32()` return a COMXX port prefixed with `\\.\` so that GDB is instructed to open `\\.\COM15` for example. As this works on Windows 10.

Sure thing! Can you propose it in the form of a pull request? :-)

fabiobaltieri commented:
> 



> I know this has been closed. However I'm getting
> 
> ```
> COM15: No such file or directory.
> "monitor" command not supported by this target.
> Don't know how to attach.  Try "help target".
> You can't do that when your target is `None'
> The program is not being run.
> ```
> 
> May I propose that `blackmagicprobe_gdb_serial_win32()` return a COMXX port prefixed with `\\.\` so that GDB is instructed to open `\\.\COM15` for example. As this works on Windows 10.
> 
> I'm using Zephyr SDK `v0.15.2`, and Zephyr `v3.3.0-rc1`

@TequinDragon opened https://github.com/zephyrproject-rtos/zephyr/pull/54678 for that, would you be able to verify it? (can't add you as reviewer directly)
